### ISSUE TITLE
Scheduler: optimize scheduler framework.

### ISSUE THREAD

**Issue by chaunceyjiang** [karmada-io/karmada#3318]
**What would you like to be added**:

Optimize scheduler framework.

**Why is this needed**:

When I wanted to add a scheduling plugin to the scheduler, I found it difficult to do it under the current scheduling framework. The current scheduling process lacks scheduling steps, such as: `preFilter`, `postFilter`, `preScore`, etc., and also lacks a cycle state, which is used to share data in different plug-ins.

The scheduler should only use the score to select the cluster to be scheduled, no other additional reasons are required.

So there are two goals for this optimization.
1. Increase the extension point of the scheduler
2. Convert all the information needed by the scheduler into scores.

- [ ] Optimize the scheduler's retry mechanism https://github.com/karmada-io/karmada/issues/3257
- [ ] Introduce cycle state to scheduler https://github.com/karmada-io/karmada/pull/3299
- [ ] Introduce preFilter step to scheduler
- [ ] Introduce postFilter step to scheduler
- [ ] Introduce preScore step to scheduler
- [ ] Add weights for score plugins https://github.com/karmada-io/karmada/pull/3343
- [ ] Support scheduler plugin to pass in custom parameters #3205 
- [ ] Add a scheduler score plugin based on AvailableReplicas https://github.com/karmada-io/karmada/pull/3287
- [ ] Optimize the score calculation of Provider/Region/Zone Group https://github.com/karmada-io/karmada/pull/3314
- [ ] region selection algorithm optimization https://github.com/karmada-io/karmada/pull/3259
- [ ] TODO 

Now: 
<img width="261" alt="image" src="https://user-images.githubusercontent.com/17962021/226846456-5e55062b-dc68-4ac3-938f-2f07e0ac4322.png">

Disadvantages: There are too few scheduler extension points, and data cannot be shared between plug-ins.

After:

<img width="311" alt="image" src="https://user-images.githubusercontent.com/17962021/226847072-d72463d2-cf2f-48d9-a8f4-9dd6b3a47d7b.png">

Data can be shared between different plugins via CycleState.

**Comment by XiShanYongYe-Chang** [comment]
You can add an agenda to the community meeting to introduce it in detail.

**Comment by Poor12** [comment]
I have a question, do we really need prefilter, postFilter and postScore now? Is there a definite plugin to add in this step? If we currently do not have a clear built-in plug-in at this step, how can we judge that users have the need for this extension？For example, a typical postfilter implementation is preemption, but Karmada does not involve now. We can discuss it in next meeting.

**Comment by chaunceyjiang** [comment]
Yes, that's a very good question. I'm not able to fully answer your question at the moment. As described in the issue `Convert all the information needed by the scheduler into scores`, I just wanted to convert `AvailableReplicas` into scores. However, as you can see in [the PR](https://github.com/karmada-io/karmada/pull/3287), its implementation is very poor due to insufficient extensibility of the current scheduling framework. 

Hopefully we will have a clear conclusion in Karmada next meeting.

**Comment by Garrybest** [comment]
I think `Prefilter` is necessary so far, we need this extension point to get all available replicas. Other extension could be added in the future if it is necessary.

**Comment by chaunceyjiang** [comment]
For the current implementation of the assignment algorithm based on weight, the weight information in `Placement` is directly read. I have an idea, is it possible to convert the weights into scores and assign `Replicas` according to the weight of the scores?

**Comment by Garrybest** [comment]
I have thought this before. Firstly by Filter plugin, we caculate the upper boundary replicas of every member cluster. And then Score plugin help us caculate the weight of all member clusters. But I'm not sure if this method could meet the requirement of all scenarios.

**Comment by Poor12** [comment]
I find that the scheduler uses cluster resources in the three steps of `score`, `selectClusters` and `assignReplicas`. These steps are not dependent on timing and can use the same data. 
I wonder if we can put cluster resources into the `snapshot`. We use the same data in all steps of one scheduling.
```
// Snapshot is a snapshot of cache ClusterInfo. The scheduler takes a
// snapshot at the beginning of each scheduling cycle and uses it for its operations in that cycle.
type Snapshot struct {
	// clusterInfoList is the list of nodes as ordered in the cache's nodeTree.
	clusterInfoList []*framework.ClusterInfo
}
```

**Comment by chaunceyjiang** [comment]
> We use the same data in all steps of one scheduling.


This is the core idea of my PR #3299.

**Comment by Garrybest** [comment]
I'm trying to design a new scheduelr framework. A little bit difficult. I will show you ASAP.

**Comment by jwcesign** [comment]
Is it possible to extract this framework as a utility library? Doing so would enable other areas to utilize the scheduling logic in the future, such as FederatedHPA.

**Comment by RainbowMango** [comment]
> I'm trying to design a new scheduelr framework. A little bit difficult. I will show you ASAP.

👍

**Comment by chaosi-zju** [comment]
@chaunceyjiang @Garrybest Hello, is this `new scheduler framework` task is on going? If it is temporarily interrupted, I am very interested in resuming and following up on this task. It sounds like a very challenging and interesting task and I am willing to have a try~

**Comment by Garrybest** [comment]
Hi @chaosi-zju, FYI, I have written an implementation [here](https://docs.google.com/document/d/1l6zO4xf879KdW_WPS7aMED0SUmnk487_XDsC12TtuTQ/edit#heading=h.4x1rjfh8w7mi).
